action/protocol/staking/protocol.go

```diff
@@ -114,3 +119,36 @@ func (p *Protocol) handleRestake(ctx context.Context, act action.Action, sm prot
 	// TODO
 	return nil, nil
 }
+
+func (p *Protocol) hasDelegate(name address.Address) bool {
+	return p.delegates.Contains(name)
+}
+
+func (p *Protocol) addDelegate(d *Delegate) error {
+	if err := d.Validate(); err != nil {
+		return err
+	}
+
+	if !p.hasDelegate(d.Owner) {
+		p.delegates[d.Owner.String()] = d
+	}
+	return nil
+}
+
+func (p *Protocol) delDelegate(name address.Address) {
+	delete(p.delegates, name.String())
+}
+
+func (p *Protocol) addVote(name address.Address, amount *big.Int) error {
+	if p.hasDelegate(name) {
+		return p.delegates[name.String()].AddVote(amount)
+	}
+	return nil
+}
+
+func (p *Protocol) subVote(name address.Address, amount *big.Int) error {
+	if p.hasDelegate(name) {
+		return p.delegates[name.String()].SubVote(amount)
+	}
+	return nil
```

Comment:
return some error?

Reply:
yes, need to think the correct logic and re-write this entire thing